### PR TITLE
Included tests and handled stack overflow error for parser

### PR DESCRIPTION
**Pull request by nathannaveen**
# Description of the PR

- Included tests for `parseHelper`
- Included tests for `parse`
- A graph nodes children is pointing back at itself can cause an infinite loop, resulting in a stack overflow error (i.e. root points to its child, and the child points back at the root, `root.Next = root.Child` and `root.Child.Next = root`). To solve this issue, I have included a map to store all visited nodes so that we don't keep calling nodes that have already been called. I have also included a test for a stack overflow.

# PR Checklist

- [x] All commits have [a Developer Certificate of Origin (DCO)](https://wiki.linuxfoundation.org/dco) -- they are generated using `-s` flag to `git commit`.
- [x] All new changes are covered by tests
- [ ] If GraphQL schema is changed, `make generate` has been run
- [ ] If `collectsub` protobuf has been changed, `make proto` has been run
- [ ] All CI checks are passing (tests and formatting)
- [ ] All dependent PRs have already been merged

### CODE REVIEW

File: pkg/ingestor/parser/parser_test.go [hunk 1]
```diff
@@ -15,210 +15,192 @@
...
-// 				if len(d) != len(tt.want) {
-// 					t.Errorf("ParseDocumentTree() = %v, want %v", d, tt.want)
-// 				}
-// 				for i := range d {
-// 					compare(t, d[i].Edges, tt.want[i].Edges, d[i].Nodes, tt.want[i].Nodes)
-// 				}
-// 				return nil
-// 			}
-//
-// 			err = Subscribe(ctx, transportFunc)
```
Comment (pxp928) [inline pkg/ingestor/parser/parser_test.go:201]: We should revive this test as well to test the functionality for the NATS Subscriber.

**Review by pxp928 — COMMENTED** [review]
Thanks @nathannaveen! This looks good so far.

File: pkg/ingestor/parser/internal/mock/types.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
+// Code generated by MockGen. DO NOT EDIT.
+// Source: /Users/nathannaveen/go/src/github.com/nathannaveen/guac/pkg/ingestor/parser/common/types.go
+
+// Package mocks is a generated GoMock package.
+package mocks
```
Comment (lumjjb) [inline pkg/ingestor/parser/internal/mock/types.go:5]: does this come with a configuration file? want to ensure that this gets updated when it needs to be. Should we add a make command for this? 

Also a broader question, on testing consistency, i don't think it matters too much now that we have tests that use the declaration type tests definitions vs the just code that does comparison (like current other tests). I do worry a bit on consistency and maintainability. Thoughts @pxp928 @mihaimaruseac @mihaimaruseac @jeffmendoza

**Review by lumjjb — APPROVED** [review]
minor nits and some open questions, else lgtm

**Review by lumjjb — APPROVED** [review]
minor nits and some open questions, else lgtm

**Review by lumjjb — APPROVED** [review]
minor nits and some open questions, else lgtm

File: pkg/ingestor/parser/internal/mock/types.go (hunk 2 above)
Comment (mihaimaruseac) [reply pkg/ingestor/parser/internal/mock/types.go:5]: +1 to adding to Makefile and [CI](https://github.com/guacsec/guac/blob/131eeab8164633412cf15109e7d98fab5ce46dc3/.github/workflows/ci.yaml#L96-L97) to make sure it is always regenerated as needed.

Regarding test suites: I'd recommend striving for consistency, as that makes it easier to transfer knowledge from one area of GUAC to another.

File: pkg/ingestor/parser/parser_test.go (hunk 1 above)
Comment (nathannaveen) [reply pkg/ingestor/parser/parser_test.go:201]: I don't want to make this PR too large, so I will do that in a subsequent PR.

File: pkg/ingestor/parser/internal/mock/types.go (hunk 2 above)
Comment (nathannaveen) [reply pkg/ingestor/parser/internal/mock/types.go:5]: What do you mean when you say the following? Could you please explain?
> Also a broader question, on testing consistency, i don't think it matters too much now that we have tests that use the declaration type tests definitions vs the just code that does comparison (like current other tests). I do worry a bit on consistency and maintainability. Thoughts @pxp928 @mihaimaruseac @mihaimaruseac @jeffmendoza

Comment (nathannaveen) [reply pkg/ingestor/parser/internal/mock/types.go:5]: I will add this to the CI and makefile.

File: pkg/ingestor/parser/parser.go [hunk 3]
```diff
@@ -143,22 +143,33 @@ func ParseDocumentTree(ctx context.Context, docTree processor.DocumentTree) ([]a
 	return assemblerInputs, identifierStrings, nil
 }
 
-func (t *docTreeBuilder) parse(ctx context.Context, root processor.DocumentTree) error {
+// visitedKey is used to keep track of the document nodes that have already been visited to avoid infinite loops.
+// The key is a combination of the document type, format, and source information. This is used instead of
+// processor.Document because maps cannot use slices as keys.
+type visitedKey struct {
+	Type              processor.DocumentType
```
[the remark this answers is not in the document]
Comment (nathannaveen) [reply pkg/ingestor/parser/parser.go]: You are right; I have made all of them internal.

File: pkg/ingestor/parser/parser_test.go [hunk 4]
```diff
@@ -15,210 +15,192 @@
...
+	type fields struct {
+		identities    []common.TrustInformation
+		graphBuilders []*common.GraphBuilder
+	}
+	tests := []struct {
+		name            string
+		fields          fields
+		root            processor.DocumentTree
+		registerDocType processor.DocumentType
+		makeOverflow    bool
```
[the remark this answers is not in the document]
Comment (nathannaveen) [reply pkg/ingestor/parser/parser_test.go]: 👍 I have added comments to `registerDocType` and `makeOverflow`.

File: pkg/ingestor/parser/internal/mock/types.go (hunk 2 above)
Comment (nathannaveen) [reply pkg/ingestor/parser/internal/mock/types.go:20]: I have updated the Makefile, but didn't need to update the CI because I added the mockgen to the generate which is already being called in the CI.

Comment (pxp928) [reply pkg/ingestor/parser/internal/mock/types.go:20]: > What do you mean when you say the following? Could you please explain?

So the other tests [collector](https://github.com/guacsec/guac/blob/main/pkg/handler/collector/collector_test.go) and [processor](https://github.com/guacsec/guac/blob/main/pkg/handler/processor/process/process_test.go) all do some type of comparison to check that it received the proper documents/artifacts that the results should be as part of the unit tests. This is a deviation from those so the question comes to whether we want to keep them all consistent such that it's easier to understand/maintain (and for doing knowledge transfer in the future).

File: Makefile [hunk 5]
```diff
@@ -91,6 +91,7 @@ fmt-md:
 .PHONY: generate
 generate:
 	go generate ./...
+	mockgen -source=pkg/ingestor/parser/common/types.go -destination=pkg/ingestor/parser/internal/mock/types.go -package=mocks -copyright_file=pkg/ingestor/parser/internal/mock/license.txt
```
Comment (pxp928) [inline Makefile:94]: This can be moved to a `mockgen.go` similar to what we did here for the [scorecard mock gen](https://github.com/guacsec/guac/blob/main/internal/testing/mocks/mockgen.go). So that when `go generate ./...` is called, it will run the proper mockgen.

File: pkg/ingestor/parser/internal/mock/types.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+// Code generated by MockGen. DO NOT EDIT.
+// Source: pkg/ingestor/parser/common/types.go
+
+// Package mocks is a generated GoMock package.
+package mocks
```
Comment (pxp928) [inline pkg/ingestor/parser/internal/mock/types.go:20]: Could you move this into `internal/testing` or a subfolder in `internal/testing/mocks` to keep it all in once place?

**Review by pxp928 — COMMENTED** [review]
Couple of comments! LGTM otherwise besides the open question.

File: pkg/ingestor/parser/internal/mock/types.go (hunk 2 above)
Comment (nathannaveen) [reply pkg/ingestor/parser/internal/mock/types.go]: Thanks for the explanation, Parth! 

So, what is the outcome? Should I rewrite my tests in a similar way to the standard?

File: Makefile (hunk 5 above)
Comment (nathannaveen) [reply Makefile]: 👍

File: pkg/ingestor/parser/internal/mock/types.go (hunk 6 above)
Comment (nathannaveen) [reply pkg/ingestor/parser/internal/mock/types.go]: Moved the file to `internal/testing/mocks`.

**Review by pxp928 — APPROVED** [review]
LGTM! Thanks @nathannaveen

File: .github/scripts/excluded_from_copyright [hunk 7]
```diff
@@ -49,3 +49,4 @@
 ./pkg/assembler/graphql/resolvers/path.resolvers.go
 ./pkg/assembler/graphql/generated/path.generated.go
 ./internal/testing/mocks/scorecard.go
+./internal/testing/mocks/documentparser.go
```
Comment (pxp928) [inline .github/scripts/excluded_from_copyright:52]: small nit: new line at end of file else LGTM

**Review by pxp928 — APPROVED** [review]

File: pkg/ingestor/parser/internal/mock/types.go (hunk 2 above)
Comment (nathannaveen) [reply pkg/ingestor/parser/internal/mock/types.go]: Ok, sure, I will add the "comparison" checks for the `Test_ParserSubscribe` in the next PR.

**Comment by pxp928** [comment]
@nathannaveen If you rebase on main, we fixed the CI issues. After that, it should auto-merge this PR.
